### PR TITLE
Fix GA loss bugs and add unit test

### PR DESCRIPTION
**Pull request by techkang**
# What does this PR do?
There are two ways to fix GA loss bugs:
1. Use `num_items_in_batch` in loss function defined by model. In this case, `model_accepts_loss_kwargs` is `True`.
2. The model doesn't have loss function or user has self-defined loss function, which is `compute_loss_func`.

However, previes unit test only test for the second condition. So I introduced a new unit test to cover the first condition and fix the bugs by the way.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@muellerzr @ArthurZucker

### CODE REVIEW

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -836,15 +926,16 @@ def compute_loss(logits, labels, vocab_size, num_items_in_batch, disable_num_ite
...
-        # These should be very off
-        for diff in diff_broken:
-            self.assertGreater(abs(diff), 0.1, f"Difference {diff} is not greater than 0.1")
+        diff_truth = [
+            abs(base - grad) for base, grad in zip(base_loss_callback.losses, grad_accum_loss_callback.losses)
+        ]
+        diff_broken = [abs(base - grad) for base, grad in zip(base_loss_callback.losses, broken_loss_callback.losses)]
+
+        # all diff truth should be quite close
+        self.assertLess(max(diff_truth), 0.3, f"Difference {max(diff_truth)} is not within 0.3")
```
Comment (muellerzr) [inline tests/trainer/test_trainer.py:935]: Let's be a bit more aggressive and do 0.15 (this passes). I still feel that's quite big but I can't figure out why (my tests showed 0.001 should be doable).

```suggestion
        self.assertLess(max(diff_truth), 0.15, f"Difference {max(diff_truth)} is not within 0.15")
```

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -750,7 +750,97 @@ def test_model_init(self):
...
+        trainer.train()
+
+        # Calculate the difference between the base loss and the grad_accum loss
+        diff_truth = [
+            abs(base - grad) for base, grad in zip(base_loss_callback.losses, grad_accum_loss_callback.losses)
+        ]
+        diff_broken = [abs(base - grad) for base, grad in zip(base_loss_callback.losses, broken_loss_callback.losses)]
+
+        # all diff truth should be quite close
+        self.assertLess(max(diff_truth), 0.3, f"Difference {max(diff_truth)} is not within 0.3")
```
Comment (muellerzr) [inline tests/trainer/test_trainer.py:837]:
```suggestion
        self.assertLess(max(diff_truth), 0.2, f"Difference {max(diff_truth)} is not within 0.2")
```

Similarly we can be aggressive here too

**Review by muellerzr — APPROVED** [review]
Thanks! These solutions make sense and ran the tests myself. Left some nits for less-leeway on the test closeness. cc @ArthurZucker

File: tests/trainer/test_trainer.py (hunk 1 above)
Comment (techkang) [reply tests/trainer/test_trainer.py:935]: It is strange that I tested on both Mac and Windows that `max(diff_truth)` is 0.144. So maybe 0.15 may failed on some other machine.

File: tests/trainer/test_trainer.py (hunk 2 above)
Comment (techkang) [reply tests/trainer/test_trainer.py:837]: `max(diff_truth)` is 0.1997 on my machine and I think 0.2 is too aggressive.

Comment (techkang) [reply tests/trainer/test_trainer.py]: I managed to reduce the gap to 1e-4 by padding all input labels to the same length. However, this method did not work for the GPT-2 model. I will continue to explore other solutions.

**Review by ArthurZucker — COMMENTED** [review]
Sorry @muellerzr but this does not solve:
```
FAILED examples/pytorch/test_pytorch_examples.py::ExamplesTests::test_run_speech_recognition_seq2seq - TypeError: Wav2Vec2Model.forward() got an unexpected keyword argument 'num_items_in_batch'
```
so I am not sure I understand. Related to #35113 and #35128. 
We can't merge with the broken test

**Comment by techkang** [comment]
@ArthurZucker The Wav2Vec2Model bug is because SpeechEncoderDecoderModel takes variable argument as forward paramters:
https://github.com/huggingface/transformers/blob/c8c8dffbe45ebef0a8dba4a51024e5e5e498596b/src/transformers/models/speech_encoder_decoder/modeling_speech_encoder_decoder.py#L457
But it dispatches the paramater to it's encoder and decode which doesn't accept variable argument: https://github.com/huggingface/transformers/blob/c8c8dffbe45ebef0a8dba4a51024e5e5e498596b/src/transformers/models/speech_encoder_decoder/modeling_speech_encoder_decoder.py#L489-L493
I think the better solution is to modify it's decode to accept variable argument. I proposed a new commit and the test succeed.

File: tests/trainer/test_trainer.py (hunk 1 above)
Comment (techkang) [reply tests/trainer/test_trainer.py]: Done! I use TinyStories to narrow down gap to the same as you. The code is submitted.

**Comment by techkang** [comment]
Finally almost all test passed. The last not passed one I don't think is caused by this PR.

**Comment by techkang** [comment]
Finally all unit test passed. Please check  again. @muellerzr @ArthurZucker

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -750,11 +750,102 @@ def test_model_init(self):
         self.check_trained_model(trainer.model, alternate_seed=True)
 
     @slow
-    def test_gradient_accumulation_loss_alignment(self):
+    def test_gradient_accumulation_loss_alignment_with_model_loss(self):
         set_seed(42)
         import datasets
 
-        model_name = "distilgpt2"
+        model_name = "nickypro/tinyllama-110M"
```
Comment (ArthurZucker) [inline tests/trainer/test_trainer.py:757]: Okay, would be nice if we had a safetensors model here but alright.

**Review by ArthurZucker — APPROVED** [review]
Thanks a lot @techkang ! 
I did not dive enough on the test, my bad 🤗 
Merging ASAP and doing the patch

**Pull request merged**